### PR TITLE
Initial implementation of RLE

### PR DESCRIPTION
**Pull request by rhattersley**

### CODE REVIEW

File: lib/mo_pack/_packing.pyx [hunk 1]
```diff
@@ -65,6 +72,31 @@ cdef void MO_syslog(int value, char* message, const function* const caller):
...
+    """
+    Unpack the given data using run-length encoding.
+
+    """
+    cdef np.ndarray[np.float32_t, ndim=2, mode="c"] unpacked_data
+    unpacked_data = np.empty([nrows, ncols], dtype=np.float32)
+
+    cdef np.ndarray[np.float32_t, ndim=1] packed_data
+    packed_data = np.frombuffer(packed_data_buffer, dtype=np.float32)
+    packed_data = packed_data.byteswap()
```
Comment (rhattersley) [inline lib/mo_pack/_packing.pyx]: Might need to be smarter about where/when we do this!

File: lib/mo_pack/_packing.pyx [hunk 2]
```diff
@@ -65,6 +71,33 @@ cdef void MO_syslog(int value, char* message, const function* const caller):
 #    printf('\n')
 
 
+def unpack_rle(packed_data_buffer, int nrows, int ncols,
```
Comment (lbdreyer) [inline lib/mo_pack/_packing.pyx]: Why did you choose to call it "unpack_rle" rather than "decode_rle"?

Comment (rhattersley) [inline lib/mo_pack/_packing.pyx]: To make it consistent with `unpack_wgdos`.

**Comment by lbdreyer** [comment]
Great, this just needs a test then it should mergeable.

**Comment by pelson** [comment]
Agree. A test is important. Incidentally, it might be easier to test if you implement pack_rle / or encode_rle.

In terms of consistency, I think in this case, the term "pack" isn't quite as applicable - I'd not be averse to calling it something different (encode / decode) / (compress / decompress) etc.

**Comment by rhattersley** [comment]
> I'd not be averse to calling it something different (encode / decode) / (compress / decompress) etc.

Calling just the RLE routines something different? Or both the RLE and WGDOS routines?

I don't like having to remember that it's `unpack` for WGDOS but `decode` for RLE. But I'm happy to have `[de]compress_(rle|wgdos)`, `(en|de)code_(rle|wgdos)`, etc.

**Comment by rhattersley** [comment]
Changes since last time:
 - I've added support for RLE compression. This uses a *fork* of the routine in libmo_unpack to avoid a [buffer overflow](https://en.wikipedia.org/wiki/Buffer_overflow) bug in the original. We will need to push this fix upstream (see SciTools/libmo_unpack#2) ASAP, but I didn't want to get mo_pack bogged down doing that as the processes for making a new libmo_unpack release aren't clear yet.
 - The routines have been renamed as: `compress_rle`, `compress_wgdos`, `decompress_rle`, `decompress_wgdos`.

**Comment by rhattersley** [comment]
> Changes since last time:

- And I've added a bunch of tests. :smile:

**Comment by rhattersley** [comment]
> We will need to push this fix upstream (see SciTools/libmo_unpack#2) ASAP

This has been fixed upstream now, but would still need a release (which might require its build process to be cleaned up first).

**Comment by lbdreyer** [comment]
Should this wait for the libmo_unpack release? (https://github.com/SciTools/libmo_unpack/pull/14)

**Comment by lbdreyer** [comment]
If we decide that mo_pack 0.2.0 depends on libmo_unpack 0.3.0 then we will also need to update the libmo_unpack conda recipe

**Comment by rhattersley** [comment]
> Should this wait for the libmo_unpack release?

Yes - I think that's probably for the best. Then I can remove the forked code.

**Comment by rhattersley** [comment]
> If we decide that mo_pack 0.2.0 depends on libmo_unpack 3.0 then we will also need to update the libmo_unpack conda recipe

See SciTools/conda-recipes-scitools#85 :smile:

**Comment by rhattersley** [comment]
libmo_unpack 3.0 has been [released](https://github.com/SciTools/libmo_unpack/releases/tag/v3.0) and the [corresponding conda package](https://github.com/SciTools/conda-recipes-scitools/pull/85) has been [uploaded to anaconda.org](https://anaconda.org/SciTools/libmo_unpack/files) ... so, I've removed the RLE fix from here. I think this is ready to merge now...

**Comment by rhattersley** [comment]
> I think this is ready to merge now...

Although we might want to squash the commits first.

File: lib/mo_pack/tests/test_rle.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+# (at your option) any later version.
+#
+# mo_pack is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with mo_pack. If not, see <http://www.gnu.org/licenses/>.
+"""Tests for the `mo_pack.compress_rle` function."""
```
Comment (lbdreyer) [inline lib/mo_pack/tests/test_rle.py]: You are testing both compress_rle and decompress_rle here.

**Comment by lbdreyer** [comment]
I would change that docstring but otherwise all good!

**Comment by rhattersley** [comment]
Thanks @lbdreyer - all fixed and squashed.

**Pull request merged**